### PR TITLE
Convert file protocol error handling to failure

### PR DESCRIPTION
**Pull request by plauche**
Crates to convert to failure:
- [x] cbor_protocol
- [x] file_protocol
- [x] file_client
- [x] file_service

### CODE REVIEW

**Comment by plauche** [comment]
@CatGarab I need some help on naming here :black_flag:! I'm also not sure if we want more granular errors or if we're ok with embedding descriptive strings.

**Comment by plauche** [comment]
[File service: error handling conversion](https://trello.com/c/FMlhZ9Ks/67-file-service-error-handling-conversion)

File: libs/cbor-protocol/src/lib.rs [hunk 1]
```diff
@@ -224,14 +262,14 @@ impl Protocol {
...
+    pub fn recv_message(&self) -> Result<serde_cbor::Value, ProtocolError> {
         let mut buf = [0; MSG_SIZE];
         let (size, _peer) = self
             .handle
             .recv_from(&mut buf)
-            .map_err(|err| format!("Failed to receive a message: {}", err))?;
+            .map_err(|err| ProtocolError::ReceiveFailed { err })?;
 
-        self.recv_start(&buf[0..size]).map_err(|err| err)
+        Ok(self.recv_start(&buf[0..size])?)
```
Comment (CatGarab) [inline libs/cbor-protocol/src/lib.rs]: What's the reason for wrapping this in `Ok()`?
Is there a difference between these two?
`Ok(self.recv_start(&buf[0..size])?)`
vs
`self.recv_start(&buf[0..size])`

File: libs/file-protocol/src/lib.rs [hunk 2]
```diff
@@ -85,7 +84,7 @@
 //! ```
 //!
 
-#![deny(missing_docs)]
+// #![deny(missing_docs)]
```
Comment (CatGarab) [inline libs/file-protocol/src/lib.rs]: nit: uncomment this

File: libs/file-protocol/src/lib.rs [hunk 3]
```diff
@@ -95,6 +94,8 @@ extern crate rand;
 extern crate serde;
 extern crate serde_cbor;
 extern crate time;
+#[macro_use]
+extern crate failure;
```
Comment (CatGarab) [inline libs/file-protocol/src/lib.rs]: nit: not in alphabetical order

File: libs/file-protocol/src/lib.rs [hunk 4]
```diff
@@ -108,6 +109,59 @@ pub use parsers::parse_channel_id;
...
+        err: serde_cbor::error::Error,
+    },
+    #[fail(display = "Storage failed to {}: {}", action, err)]
+    Storage { action: String, err: std::io::Error },
+    #[fail(display = "Failed to serialize: {}", err)]
+    Serialize { err: serde_cbor::error::Error },
+    #[fail(display = "File hash mismatch")]
+    HashMismatch,
+    #[fail(display = "Failed to finalize file: {}", cause)]
+    FinializeFailed { cause: String },
```
Comment (CatGarab) [inline libs/file-protocol/src/lib.rs]: nit: name tweak to match the other error names. FinalizeFailed -> FinalizeError

File: libs/file-protocol/src/storage.rs [hunk 5]
```diff
@@ -33,89 +34,110 @@ use time;
...
-    fs::create_dir_all(file_dir.clone())
-        .map_err(|err| format!("Failed to create temp storage directory: {:?}", err))?;
+    fs::create_dir_all(file_dir.clone()).map_err(|err| ProtocolError::Storage {
+        action: "create temp storage directory".to_owned(),
+        err,
+    })?;
 
     let meta_path = file_dir.join("meta");
-    let temp_path = file_dir.join(".meta.tmp");
+    let temp_path = file_dir.join(".meta.mp");
```
Comment (CatGarab) [inline libs/file-protocol/src/storage.rs]: why the extension name change?

File: libs/file-protocol/src/storage.rs [hunk 6]
```diff
@@ -218,44 +245,55 @@ pub fn validate_file(
...
+                let chunk = reader.fill_buf().map_err(|err| ProtocolError::IoError {
+                    action: format!("read chunk from source"),
+                    err,
+                })?;
                 if chunk.len() == 0 {
-                    output
-                        .sync_all()
-                        .map_err(|err| format!("Failed to sync {:?}: {}", temp_path, err))?;
+                    output.sync_all().map_err(|err| ProtocolError::IoError {
+                        action: format!("Failed to sync {:?}", temp_path),
```
Comment (CatGarab) [inline libs/file-protocol/src/storage.rs]: nit: You've made all of the other action messages start with a lowercase letter

File: libs/file-protocol/src/storage.rs [hunk 7]
```diff
@@ -134,8 +153,10 @@ pub fn load_meta(prefix: &str, hash: &str) -> Result<u32, String> {
                         None
                     }
                 })
-        })
-        .ok_or("Failed to parse temporary file's metadata".to_owned())?;
+        }).ok_or(ProtocolError::StorageParseError(
+            "Failed to parse temporary file's metadata".to_owned(),
+        ))?;
+    //.ok_or("Failed to parse temporary file's metadata".to_owned())?;
```
Comment (CatGarab) [inline libs/file-protocol/src/storage.rs]: nit: debug line

File: libs/file-protocol/src/storage.rs [hunk 8]
```diff
@@ -33,89 +33,108 @@ use time;
...
 
     File::open(meta_path)
-        .map_err(|err| format!("Unable to open {} metadata file: {}", hash, err))?
-        .read_to_end(&mut data)
-        .map_err(|err| format!("Unable to read {} metadata file: {}", hash, err))?;
-
-    let metadata: Value = de::from_slice(&data)
-        .map_err(|err| format!("Unable to parse metadata for {}: {}", hash, err))?;
+        .map_err(|err| ProtocolError::StorageError {
+            action: format!("Unable to open {} metadata file", hash),
```
Comment (CatGarab) [inline libs/file-protocol/src/storage.rs]: nit: This wording doesn't match what you've got in the other action messages

File: libs/file-protocol/src/storage.rs [hunk 9]
```diff
@@ -159,8 +180,10 @@ pub fn validate_file(
 
     let mut prev_entry: i32 = -1;
 
-    let entries = fs::read_dir(hash_path.clone())
-        .map_err(|err| format!("Failed to read {:?} directory: {}", hash_path, err))?;
+    let entries = fs::read_dir(hash_path.clone()).map_err(|err| ProtocolError::StorageError {
+        action: format!("Failed to read {:?} directory", hash_path),
```
Comment (CatGarab) [inline libs/file-protocol/src/storage.rs]: nit: doesn't match other action message wording

**Pull request merged**